lib/action_client/middleware/parser.rb

```diff
@@ -0,0 +1,61 @@
+module ActionClient
+  module Middleware
+    class Parser
+      JsonParser = proc do |body|
+        JSON.parse(body, object_class: HashWithIndifferentAccess)
+      end
+      XmlParser = proc do |body|
+        Nokogiri::XML.parse(body).tap do |document|
+          document.validate
+
+          document.errors.each do |error|
+            if error.is_a?(Nokogiri::XML::SyntaxError)
+              raise error
+            end
+          end
+        end
+      end
+      NullParser = -> (body) { body }
+
+      class_attribute :parsers, default: {
+        "application/json": JsonParser,
+        "application/xml": XmlParser,
+      }.with_indifferent_access
+
+      def initialize(app, configuration = {})
+        @app = app
+        @parsers = self.class.parsers.merge(configuration.fetch(:parsers, {}))
+      end
+
+      def call(env)
+        status, headers, body_proxy = @app.call(env)
+        body = body_proxy.each(&:yield_self).join
+        content_type = headers[Rack::CONTENT_TYPE].to_s
+
+        if body.present?
+          parser = fetch_parser_for_content_type(content_type)
+
+          begin
+            [ status, headers, parser.call(body) ]
+          rescue StandardError => error
+            raise ActionClient::ParseError.new(error, body, content_type)
```

Comment:
Something like [Python's exception chaining][1] would be nice here: it lets you raise a new exception explicitly specifying another exception as the cause.

I don't think Ruby has anything similar, so passing the original exception as an argument is probably as near as we'll get.

[1]: https://www.python.org/dev/peps/pep-3134/